src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/HighHeapUsageClusterRca.java

```diff
@@ -48,133 +42,91 @@
+    long timestamp = clock.millis();
+    // for each RCA type this cluster RCA subscribes, read its most recent flowunit and if it is
+    // unhealthy, append this flowunit to output node summary
+    for (Rca<ResourceFlowUnit<HotNodeSummary>> nodeRca : this.nodeRcas) {
+      // skip if we haven't receive any flowunit from this RCA yet.
+      if (nodeTable.get(nodeKey, nodeRca.name()) == null) {
         continue;
       }
-      String nodeId = hotNodeRcaFlowUnit.getSummary().getNodeID().toString();
-      try {
-        readComputeWrite(nodeId, hotNodeRcaFlowUnit);
-      } catch (ExecutionException e) {
-        LOG.debug("ExecutionException occurs when retrieving key {}", nodeId);
+      List<ResourceFlowUnit<HotNodeSummary>> flowUnits =
+          ImmutableList.copyOf(nodeTable.get(nodeKey, nodeRca.name()));
+      List<HotResourceSummary> oldGenSummaries = new ArrayList<>();
+      List<HotResourceSummary> youngGenSummaries = new ArrayList<>();
+      for (ResourceFlowUnit<HotNodeSummary> flowUnit : flowUnits) {
+        // skip this flowunit if :
+        // 1. the timestamp of this flowunit expires
+        // 2. flowunit is healthy
+        // 3. flowunit does not have summary attached to it
+        if (timestamp - flowUnit.getTimeStamp() > expirationTimeWindow
+            || flowUnit.getResourceContext().isHealthy()
+            || flowUnit.getSummary() == null) {
+          continue;
+        }
+        if (flowUnit.getResourceContext().getState() == Resources.State.UNHEALTHY) {
+          HotNodeSummary currentNodeSummary = flowUnit.getSummary();
+          for (HotResourceSummary resourceSummary : currentNodeSummary.getHotResourceSummaryList()) {
+            if (resourceSummary.getResource().getResourceEnum() == ResourceEnum.YOUNG_GEN) {
+              youngGenSummaries.add(resourceSummary);
+            }
+            else if (resourceSummary.getResource().getResourceEnum() == ResourceEnum.OLD_GEN) {
+              oldGenSummaries.add(resourceSummary);
+            }
+          }
+        }
       }
-    }
-    if (counter == rcaPeriod) {
-      List<HotNodeSummary> unhealthyNodeList = getUnhealthyNodeList();
-      counter = 0;
-      ResourceContext context = null;
-      HotClusterSummary summary = null;
-      LOG.debug("Unhealthy node id list : {}", unhealthyNodeList);
-      if (unhealthyNodeList.size() > 0) {
-        context = new ResourceContext(Resources.State.UNHEALTHY);
-        summary = new HotClusterSummary(
-            getAllClusterInstances().size(),
-            unhealthyNodeList.size());
-        for (HotNodeSummary unhealthyNodeSummary : unhealthyNodeList) {
-          summary.appendNestedSummary(unhealthyNodeSummary);
+      // youngGenSummaries can have multiple elements but we will only consider it as unhealthy if
+      // three consecutive summaries are all unhealthy and we will then pick the first element as the summary for output.
+      if (youngGenSummaries.size() >= UNHEALTHY_FLOWUNIT_THRESHOLD
```

Comment:
This actually comes from legacy code and I think it might be unnecessary to keep track of most recent three flowunits anymore. Let's keep this JVM cluster RCA simple by only keeping track of single flowunit form each node

Reply:
Fixed